documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1182 @@
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>, IComparisonOperators<T, T, bool>;
+
+    /// <summary>Subtracts each element in a second quaternion from its corresponding element in a first quaternion.</summary>
+    /// <param name="value1">The first quaternion.</param>
+    /// <param name="value2">The second quaternion.</param>
+    /// <returns>The quaternion containing the values that result from subtracting each element in <paramref name="value2" /> from its corresponding element in <paramref name="value1" />.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static Quaternion<T> Subtract<T>(Quaternion<T> value1, Quaternion<T> value2)
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+    
+    /// <summary>Calculates the length of the quaternion.</summary>
+    /// <returns>The computed length of the quaternion.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static T Length<T>(this Quaternion<T> self) where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+
+    /// <summary>Calculates the squared length of the quaternion.</summary>
+    /// <returns>The length squared of the quaternion.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static T LengthSquared<T>(this Quaternion<T> self) where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+}
+```
+
+# Geometric Types
+
+The following Geometric Types are defined:
+- Box2DF
+- Box2DI
+- Box3DF
+- Box3DI
+- CircleF
+- CircleI
+- CubeF
+- CubeI
+- PlaneF
+- PlaneI
+- Ray2DF
+- Ray2DI
+- Ray3DF
+- Ray3DI
+- RectangleF
+- RectangleI
+- SphereF
+- SphereI
+
+The Box structs are defined by Min and Max Vectors, while Rectangle and Cube are defined by Origin and Size Vectors. These Types should be implicitly castable between each other.
+
+Each type should include the following:
+- Intersect functions with both another instance of the type and a point
+- GetDistanceToNearest(Point,Edge,etc) functions should be available for a given point
+- For all but the rays and planes, GetInflated function that takes a point and returns the scaled object that is closest to the original and contains the given point
+- Include Scale and Translation transformation functions
+- For Planes and Rays, Normalize functions
+- For Planes include the following static functions
+  - CreateFromVerticies
+  - Dot
+    - with a Vector4
+  - DotCoordinate and DotNormal
+    - with a Vector3
+  - Transform
+    - With a Matrix4x4 or Quaternion, if relevant
+
+# Vector Interface Definitions
+
+The following interfaces are defined for Vectors:
+
+```cs
+public interface IVector
+{
+    /// <summary>
+    /// The type of scalars in this vector type.
+    /// </summary>
+    protected static abstract Type ScalarType { get; }
+    
+    /// <summary>
+    /// The amount of scalars in this vector type.
+    /// </summary>
+    protected internal static abstract int Count { get; }
```

Comment:
I think this has to be changed due to naming conflict with IReadOnlyList.Count.

Reply:
Could rename to Extent?

Reply:
Dimensions would probably be a better name here.

Reply:
This makes sense, I'm updating to add this to the requirements